pkg/export/transform_test.go

```diff
@@ -1193,6 +1205,200 @@ func TestSampleBuilder(t *testing.T) {
+				"metric1":         {Type: textparse.MetricTypeHistogram, Help: "metric1 help text"},
+				"metric1_a_count": {Type: textparse.MetricTypeGauge, Help: "metric1_a_count help text"},
+			}),
+			series: seriesMap{
+				1: labels.FromStrings("job", "job1", "instance", "instance1", "__name__", "metric1_sum"),
+				2: labels.FromStrings("job", "job1", "instance", "instance1", "__name__", "metric1_count"),
+				3: labels.FromStrings("job", "job1", "instance", "instance1", "__name__", "metric1_bucket", "le", "0.1"),
+				4: labels.FromStrings("job", "job1", "instance", "instance1", "__name__", "metric1_bucket", "le", "0.5"),
+				5: labels.FromStrings("job", "job1", "instance", "instance1", "__name__", "metric1_bucket", "le", "1"),
+				6: labels.FromStrings("job", "job1", "instance", "instance1", "__name__", "metric1_bucket", "le", "2.5"),
+				7: labels.FromStrings("job", "job1", "instance", "instance1", "__name__", "metric1_bucket", "le", "+Inf"),
+			},
+			samples: [][]record.RefSample{
+				// First sample set, should be skipped by reset handling.
+				// The buckets must be in ascending order for an individual histogram but otherwise
+				// no order or grouping constraints apply for series of a given histogram metric.
+				{
+					{Ref: 1, T: 1000, V: 55.1}, // hist1, sum
+					{Ref: 3, T: 1000, V: 2},    // hist1, 0.1
+					{Ref: 4, T: 1000, V: 5},    // hist1, 0.5
+					{Ref: 5, T: 1000, V: 6},    // hist1, 1
+					{Ref: 6, T: 1000, V: 8},    // hist1, 2.5
+					{Ref: 7, T: 1000, V: 10},   // hist1, inf
+					{Ref: 2, T: 1000, V: 10},   // hist1, count
+				},
+				// Second sample set should actually be emitted.
+				{
+					// Second samples for histograms should produce a distribution.
+					{Ref: 3, T: 2000, V: 4},     // hist1, 0.1
+					{Ref: 2, T: 2000, V: 21},    // hist1, count
+					{Ref: 1, T: 2000, V: 123.4}, // hist1, sum
+					{Ref: 4, T: 2000, V: 9},     // hist1, 0.5
+					{Ref: 5, T: 2000, V: 11},    // hist1, 1
+					{Ref: 6, T: 2000, V: 15},    // hist1, 2.5
+					{Ref: 7, T: 2000, V: 21},    // hist1, inf
+				},
+			},
+			exemplars: []map[chunks.HeadSeriesRef]record.RefExemplar{
+				// first sample set is skipped by reset handling
+				{},
+				{
+					// project_id, trace_id, and span_id should be in the span context
+					// random should be in the dropped labels
+					3: {Ref: 3, T: 1500, V: .099, Labels: labels.New(
+						labels.Label{Name: "project_id", Value: "1"},
+						labels.Label{Name: "trace_id", Value: "2"},
+						labels.Label{Name: "span_id", Value: "3"},
+						labels.Label{Name: "random", Value: "4"},
+					)},
+					// project_id and trace_id should both be in dropped labels
+					// since we have no span_id to make a full span context
+					4: {Ref: 4, T: 1500, V: .4, Labels: labels.New(
+						labels.Label{Name: "project_id", Value: "1"},
+						labels.Label{Name: "trace_id", Value: "2"},
+					)},
+					// timestamp is optional
+					5: {Ref: 7, V: 2},
```

Comment:
Interesting, do you remember where you saw that specified? I knew OM [allows](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#exemplars-1) for empty labelsets on exemplars - though I don't quite understand the use-case there.

What does a timestamp-less exemplar look like on a graph? Is it acceptable to omit in the Prometheus HTTP API response? I'm curious what Prometheus does when there's no timestamp on a stored exemplar...

Reply:
so I think I misunderstood what happens when an exemplar has no timestamp -- 

OM allows it, but when prometheus ingests it, it will add the current time
https://github.com/prometheus/prometheus/blob/3cde9287a65ca7cc94060befeb3f83cdca8eeb21/scrape/scrape.go#L1622. 

Let me remove that test